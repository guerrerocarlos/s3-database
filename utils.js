const { setupRecorder } = require('nock-record')

async function nockManager(testName, specificCase, specificResult, t, callback) {
	const record = setupRecorder()
	const { completeRecording, assertScopesFinished } = await record(`${testName}.${specificCase}.${specificResult}.nock`, {
		afterRecord: function (nockDefs) {
			return nockDefs.map(def => {
				try {
					const headers = def.rawHeaders.reduce(
						(acc, curr, i, arr) => {
							if (i % 2 === 0) {
								acc[arr[i].toLowerCase()] = arr[i + 1].toLowerCase();
							}
							return acc;
						},
						{}
					);
					if (
						headers['transfer-encoding'] === 'chunked' &&
						headers['content-encoding'] === 'gzip' &&
						Array.isArray(def.response)
					) {
						def.response = JSON.parse(
							zlib.gunzipSync(Buffer.from(def.response.join(''), 'hex')).toString(
								'utf-8'
							)
						);
						def.rawHeaders = Object.entries(headers).flatMap(([key, value]) => {
							if (key === 'transfer-encoding' || key === 'content-encoding') {
								return [];
							}
							return [key, value];
						});
					}
				} catch (ex) {
					console.warn('Failed to decode response', ex);
				}
				return def;
			});

		}
	});
	let result = await callback()
	completeRecording()
	t.snapshot({ action: callback.toString(), result });
	assertScopesFinished()
	return result
}

module.exports = {
	nockManager
}
