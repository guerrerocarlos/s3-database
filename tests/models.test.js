process.env.stage = 'dev'

import { nockManager } from './utils'
import { Account, Client } from './models'

jest.setTimeout = 20000

describe('Database', () => {
    test(`Create two clients`, async () => {
        await nockManager('s3_database', `create_user`, 'success', async function () {

            let client1 = Client({ id: 'clientOne', username: 'guerrerocarlos', name: 'Carlos Guerrero' })
            let client2 = Client({ id: 'clientTwo', username: '', name: '' })

            await client1.save()
            await client2.save()

            return { client1, client2 }
        })
    })

    test(`Create two accounts`, async () => {
        await nockManager('s3_database', `create_accounts`, 'success', async function () {

            let accountA = Account({ id: 'accountA', amount: '00001', type: 'usd-savings', clientId: 'clientOne' })
            let accountB = Account({ id: 'accountB', amount: '00002', type: 'usd-savings', clientId: 'clientOne' })

            await accountA.save()
            await accountB.save()

            return { accountA, accountB }
        })
    })

    test(`Get clientOne`, async () => {
        await nockManager('s3_database', `get_clientOne`, 'success', async function () {
            let client1 = await Client.get('clientOne') // equals { id: clientOne } where id is the hashKey attribute
            return { client1 }
        })
    })

    test(`Get accountA`, async () => {
        await nockManager('s3_database', `get_accountA`, 'success', async function () {
            let accountA = await Account.get('accountA')
            return { accountA }
        })
    })

    test(`Get accounts of client 'clientOne' `, async () => {
        await nockManager('s3_database', `get_clientOne_accounts`, 'success', async function () {

            let accounts = await Account.get({ clientId: 'clientOne' })

            expect(accounts.length).toEqual(2)
            return { accounts } 
        })
    })

    test(`Update clientOne's info`, async () => {
        await nockManager('s3_database', `update_clientOne`, 'success', async function () {

            let client1 = await Client.getOne('clientOne') // equals { id: clientOne } where id is the hashKey attribute
            client1.name = "Carlos Alejandro"
            await client1.save()

            client1 = await Client.getOne('clientOne')
            expect(client1.name).toEqual("Carlos Alejandro")

            return { client1 }
        })
    })

    test(`Update clientOne's info and accounts to it`, async () => {
        await nockManager('s3_database', `update_many`, 'success', async function () {

            let accounts = await Account.get({ clientId: 'clientOne' })
            await Promise.all(accounts.map(async (account) => {
                account.amount = '1000'
                await account.save()
            }))
            return accounts
        })
    })
    test(`Get or create`, async () => {
        await nockManager('s3_database', `get_or_create`, 'success', async function () {
            let client = await Client.queryOneOrCreate({ id: '005', username: 'five', name: 'Five Integer' })
            await client.save()
            return client
        })
    })
    
})