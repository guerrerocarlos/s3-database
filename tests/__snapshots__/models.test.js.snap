// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database Create two accounts 1`] = `
Object {
  "accountA": S3ObjectInstance {
    "amount": "00001",
    "clientId": "clientOne",
    "id": "accountA",
    "type": "usd-savings",
  },
  "accountB": S3ObjectInstance {
    "amount": "00002",
    "clientId": "clientOne",
    "id": "accountB",
    "type": "usd-savings",
  },
}
`;

exports[`Database Create two clients 1`] = `
Object {
  "client1": S3ObjectInstance {
    "id": "clientOne",
    "name": "Carlos Guerrero",
    "username": "guerrerocarlos",
  },
  "client2": S3ObjectInstance {
    "id": "clientTwo",
    "name": "",
    "username": "",
  },
}
`;

exports[`Database Get accountA 1`] = `
Object {
  "accountA": Array [
    S3ObjectInstance {
      "amount": "00001",
      "clientId": "clientOne",
      "id": "accountA",
      "type": "usd-savings",
    },
  ],
}
`;

exports[`Database Get accounts of client 'clientOne'  1`] = `
Object {
  "accounts": Array [
    S3ObjectInstance {
      "amount": "00001",
      "clientId": "clientOne",
      "id": "accountA",
      "type": "usd-savings",
    },
    S3ObjectInstance {
      "amount": "00002",
      "clientId": "clientOne",
      "id": "accountB",
      "type": "usd-savings",
    },
  ],
}
`;

exports[`Database Get clientOne 1`] = `
Object {
  "client1": Array [
    S3ObjectInstance {
      "id": "clientOne",
      "name": "Carlos Guerrero",
      "username": "guerrerocarlos",
    },
  ],
}
`;

exports[`Database Update clientOne's info 1`] = `
Object {
  "client1": S3ObjectInstance {
    "id": "clientOne",
    "name": "Carlos Alejandro",
    "username": "guerrerocarlos",
  },
}
`;

exports[`Database Update clientOne's info and accounts to it 1`] = `
Array [
  S3ObjectInstance {
    "amount": "1000",
    "clientId": "clientOne",
    "id": "accountA",
    "type": "usd-savings",
  },
  S3ObjectInstance {
    "amount": "1000",
    "clientId": "clientOne",
    "id": "accountB",
    "type": "usd-savings",
  },
]
`;
